frontend/src/metabase/dashboard/actions/auto-wire-parameters/utils.ts

```diff
@@ -120,21 +121,27 @@ export function getAutoWiredMappingsForDashcards(
   }
   return targetDashcardMappings;
 }
-
+// TODO: this function should automatically calculate return type based on the
+// type of dashcard
```

Comment:
Something like this should work:

```tsx
export function getParameterMappings<Card extends DashboardCard>(
  dashcard: Card,
  parameter_id: ParameterId,
  card_id: CardId,
  target: ParameterTarget | null,
): NonNullable<Card["parameter_mappings"]> {
  const isVirtual = isVirtualDashCard(dashcard);
  const isAction = isActionDashCard(dashcard);

  let parameter_mappings: NonNullable<Card["parameter_mappings"]> =
    dashcard.parameter_mappings || [];
```


And then we don't need `as DashboardParameterMapping[]` cast in `frontend/src/metabase/dashboard/actions/auto-wire-parameters/actions.ts`.

There's going to be a problem with the `parameter_mappings.concat` call below which we can replace with
```tsx
    return [
      ...parameter_mappings,
      {
        parameter_id,
        card_id,
        target,
      },
    ];
```

Reply:
I tried many ways of calculating types, but my desired goal is to get QuestionDashcardParameterMappings if we pass QuestionDashCard - I didn't find a correct way to do it

let me check this one!

Reply:
nope :( 

<img width="1325" alt="image" src="https://github.com/metabase/metabase/assets/125459446/1a30ae0a-8920-4ca3-aeb5-d924f7e830a0">

I'm not sure from product perspective, when do we need to keep card_id and when not

I'd make improvements for it in a separate PR

Reply:
@uladzimirdev did you try last snippet from [my comment](https://github.com/metabase/metabase/pull/43191#discussion_r1630806018)?

Reply:
I didn't refresh the page, sorry

Reply:
@kamilmielnik I'll rebase this branch on top of master to use ts5, with ts4 there are other errors 😄

Reply:
I'll address it in a separate PR

Reply:
https://github.com/metabase/metabase/pull/43851